<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Tackling an NLP Problem with Naive Bayes
_Author: Matt Brems_

----

We can sketch out the data science process as follows:
1. Define the problem.
2. Obtain the data.
3. Explore the data.
4. Model the data.
5. Evaluate the model.
6. Answer the problem.

In this lab, we are going to apply a **new** modeling technique to natural language processing data.

> "But how can we apply a modeling technique we haven't learned?!"

The DSI program is great - but we can't teach you *everything* about data science in 12 weeks! This lab is designed to help you start learning something new without it being taught in a formal lesson. 
- Later in the cohort (like for your capstone!), you'll be exploring models, libraries, and resources that you haven't been explicitly taught.
- After the program, you'll want to continue developing your skills. Being comfortable with documentation and being confident in your ability to read something new and decide whether or not it is an appropriate method for the problem you're trying to solve is **incredibly** valuable.

### Step 1: Define the problem.

Many organizations have a substantial interest in classifying users of their product into groups. Some examples:
- A company that serves as a marketplace may want to predict who is likely to purchase a certain type of product on their platform, like books, cars, or food.
- An application developer may want to identify which individuals are willing to pay money for "bonus features" or to upgrade their app.
- A social media organization may want to identify who generates the highest rate of content that later goes "viral."

### Summary
In this lab, you're an engineer for Facebook. In recent years, the organization Cambridge Analytica gained worldwide notoriety for its use of Facebook data in an attempt to sway electoral outcomes.

Cambridge Analytica, an organization staffed with lots of Ph.D. researchers, used the Big5 personality groupings (also called OCEAN) to group people into one of 32 different groups.
- The five qualities measured by this personality assessment are:
    - **O**penness
    - **C**onscientiousness
    - **E**xtroversion
    - **A**greeableness
    - **N**euroticism
- Each person could be classified as "Yes" or "No" for each of the five qualities.
- This makes for 32 different potential combinations of qualities. ($2^5 = 32$)
- You don't have to check it out, but if you want to learn more about this personality assessment, head to [the Wikipedia page](https://en.wikipedia.org/wiki/Big_Five_personality_traits).
- There's also [a short (3-4 pages) academic paper describing part of this approach](./celli-al_wcpr13.pdf).

Cambridge Analytica's methodology was, roughly, the following:
- Gather a large amount of data from Facebook.
- Use this data to predict an individual's Big5 personality "grouping."
- Design political advertisements that would be particularly effective to that particular "grouping." (For example, are certain advertisements particularly effective toward people with specific personality traits?)

You want to know the **real-world problem**: "Is what Cambridge Analytica attempted to do actually possible, or is it junk science?"

However, we'll solve the related **data science problem**: "Are one's Facebook statuses predictive of whether or not one is agreeable?"
> Note: If Facebook statuses aren't predictive of one being agreeable (one of the OCEAN qualities), then Cambridge Analytica's approach won't work very well!

### Step 2: Obtain the data.

Obviously, there are plenty of opportunities to discuss the ethics surrounding this particular issue... so let's do that.

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('./mypersonality_final.csv', encoding = 'ISO-8859-1')

In [3]:
data.head()

#AUTHID  \
0  b7b7764cfa1c523e4e93ab2a79a946c4   
1  b7b7764cfa1c523e4e93ab2a79a946c4   
2  b7b7764cfa1c523e4e93ab2a79a946c4   
3  b7b7764cfa1c523e4e93ab2a79a946c4   
4  b7b7764cfa1c523e4e93ab2a79a946c4   

                                              STATUS  sEXT  sNEU  sAGR  sCON  \
0                        likes the sound of thunder.  2.65   3.0  3.15  3.25   
1  is so sleepy it's not even funny that's she ca...  2.65   3.0  3.15  3.25   
2  is sore and wants the knot of muscles at the b...  2.65   3.0  3.15  3.25   
3         likes how the day sounds in this new song.  2.65   3.0  3.15  3.25   
4                                        is home. <3  2.65   3.0  3.15  3.25   

   sOPN cEXT cNEU cAGR cCON cOPN               DATE  NETWORKSIZE  BETWEENNESS  \
0   4.4    n    y    n    n    y  06/19/09 03:21 PM        180.0      14861.6   
1   4.4    n    y    n    n    y  07/02/09 08:41 AM        180.0      14861.6   
2   4.4    n    y    n    n    y  06/15/09 01:15 PM        180.0      14861.6   
3   4.4    n    y    n    n    y  06/22/09 04:48 AM        180.0      14861.6   
4   4.4    n    y    n    n    y  07/20/09 02:31 AM        180.0      14861.6   

   NBETWEENNESS  DENSITY  BROKERAGE  NBROKERAGE  TRANSITIVITY  
0         93.29     0.03    15661.0        0.49           0.1  
1         93.29     0.03    15661.0        0.49           0.1  
2         93.29     0.03    15661.0        0.49           0.1  
3         93.29     0.03    15661.0        0.49           0.1  
4         93.29     0.03    15661.0        0.49           0.1

**1. What is the difference between anonymity and confidentiality? All else held equal, which tends to keep people safer?**

Anonymity means you don't know who the participants are, while confidentiality means you know who they are but remove identifying information from your research report. Both are important ethical considerations.

Anonymity will keep people safer than confidentiality.

**2. Suppose that the "unique identifier" in the above data, the `#AUTHID`, is a randomly generated key so that it can never be connected back to the original poster. Have we guaranteed anonymity here? Why or why not?**

In this case, the anonymity can be guaranteed in certain level, but random might be pseudo-random and has a pattern that could be traced.

**3. As an engineer for Facebook, you recognize that user data will be used by Facebook and by other organizations - that won't change. However, what are at least three recommendations you would bring to your manager to improve how data is used and shared? Be as specific as you can.**

- make automatic anomity algorithm when importing/exporting data for confidentiality
- sharing must be only in the working team.
- for more security, manager will hold the password for sharing data and must enter password everytime before sharing.

### Step 3: Explore the data.

- Note: For our $X$ variable, we will only use the `STATUS` variable. For our $Y$ variable, we will only use the `cAGR` variable.

**4. Explore the data here.**
> We aren't explicitly asking you to do specific EDA here, but what EDA would you generally do with this data? Do the EDA you usually would, especially if you know what the goal of this analysis is.

In [4]:
 data.info(), data.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9917 entries, 0 to 9916
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   #AUTHID       9917 non-null   object 
 1   STATUS        9917 non-null   object 
 2   sEXT          9917 non-null   float64
 3   sNEU          9917 non-null   float64
 4   sAGR          9917 non-null   float64
 5   sCON          9917 non-null   float64
 6   sOPN          9917 non-null   float64
 7   cEXT          9917 non-null   object 
 8   cNEU          9917 non-null   object 
 9   cAGR          9917 non-null   object 
 10  cCON          9917 non-null   object 
 11  cOPN          9917 non-null   object 
 12  DATE          9917 non-null   object 
 13  NETWORKSIZE   9917 non-null   float64
 14  BETWEENNESS   9917 non-null   float64
 15  NBETWEENNESS  9917 non-null   float64
 16  DENSITY       9917 non-null   float64
 17  BROKERAGE     9917 non-null   float64
 18  NBROKERAGE    9917 non-null 

(None,
 #AUTHID         0
 STATUS          0
 sEXT            0
 sNEU            0
 sAGR            0
 sCON            0
 sOPN            0
 cEXT            0
 cNEU            0
 cAGR            0
 cCON            0
 cOPN            0
 DATE            0
 NETWORKSIZE     0
 BETWEENNESS     0
 NBETWEENNESS    0
 DENSITY         0
 BROKERAGE       0
 NBROKERAGE      0
 TRANSITIVITY    1
 dtype: int64)

In [5]:
data.dropna(inplace=True)

**5. What is the difference between CountVectorizer and TFIDFVectorizer?**

- CountVectorizer counts the number of times a word appears in the document which results in biasing in favour of most frequent words. this ends up in ignoring rare words which could have helped is in processing our data more efficiently.

    - To overcome this , we use TfidfVectorizer

- TfidfVectorizer consider overall document weightage of a word. It helps us in dealing with most frequent words. Using it we can penalize them. TfidfVectorizer weights the word counts by a measure of how often they appear in the documents.

**6. What are stopwords?**

Stop words are the words in a stop list which are filtered out because they are insignificant.

**7. Give an example of when you might remove stopwords.**

I might remove the low-level information from the text in order to give more focus to the important information

**8. Give an example of when you might keep stopwords in your model.**

I'll keep them if I want to see how it affects model with and without them.

### Step 4: Model the data.

We are going to fit two types of models: a logistic regression and [a Naive Bayes classifier](https://scikit-learn.org/stable/modules/naive_bayes.html).

**Reminder:** We will only use the feature `STATUS` to model `cAGR`.

### We want to attempt to fit our models on sixteen sets of features:

1. CountVectorizer with 100 features, with English stopwords removed and with an `ngram_range` that includes 1 and 2.
2. CountVectorizer with 100 features, with English stopwords removed and with the default `ngram_range`.
3. CountVectorizer with 100 features, with English stopwords kept in and with an `ngram_range` that includes 1 and 2.
4. CountVectorizer with 100 features, with English stopwords kept in and with the default `ngram_range`.
5. CountVectorizer with 500 features, with English stopwords removed and with an `ngram_range` that includes 1 and 2.
6. CountVectorizer with 500 features, with English stopwords removed and with the default `ngram_range`.
7. CountVectorizer with 500 features, with English stopwords kept in and with an `ngram_range` that includes 1 and 2.
8. CountVectorizer with 500 features, with English stopwords kept in and with the default `ngram_range`.
9. TFIDFVectorizer with 100 features, with English stopwords removed and with an `ngram_range` that includes 1 and 2.
10. TFIDFVectorizer with 100 features, with English stopwords removed and with the default `ngram_range`.
11. TFIDFVectorizer with 100 features, with English stopwords kept in and with an `ngram_range` that includes 1 and 2.
12. TFIDFVectorizer with 100 features, with English stopwords kept in and with the default `ngram_range`.
13. TFIDFVectorizer with 500 features, with English stopwords removed and with an `ngram_range` that includes 1 and 2.
14. TFIDFVectorizer with 500 features, with English stopwords removed and with the default `ngram_range`.
15. TFIDFVectorizer with 500 features, with English stopwords kept in and with an `ngram_range` that includes 1 and 2.
16. TFIDFVectorizer with 500 features, with English stopwords kept in and with the default `ngram_range`.

**9. Rather than manually instantiating 16 different vectorizers, what `sklearn` class have we learned about that might make this easier? Use it.**

In [6]:
from sklearn.pipeline        import Pipeline
from sklearn.linear_model    import LogisticRegression
from sklearn.naive_bayes     import MultinomialNB
from sklearn.model_selection import train_test_split, GridSearchCV

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

In [7]:
X = data['STATUS']
y = data['cAGR']

X_train,X_test,y_train,y_test = \
train_test_split(X,y,test_size=0.3,random_state=42,stratify=y)

In [8]:
pipe1 = Pipeline([('cv', CountVectorizer()),('lr', LogisticRegression())])

pipe1_params = {'cv__max_features': [100,500],
                'cv__stop_words': ['english'],
                'cv__ngram_range': [(1,1),(1,2)],
              } 

grid1 = GridSearchCV(pipe1,pipe1_params,cv=5)
grid1.fit(X_train, y_train)
print(f'CountVectorizer+LogisticRegression: {grid1.best_score_}')
# ==========================================================================
pipe2 = Pipeline([('cv', CountVectorizer()),('mb', MultinomialNB())])

pipe2_params = {'cv__max_features': [100,500],
                'cv__stop_words': ['english'],
                'cv__ngram_range': [(1,1),(1,2)],
              } 

grid2 = GridSearchCV(pipe2,pipe2_params,cv=5)
grid2.fit(X_train, y_train)
print(f'CountVectorizer+NaieveBayes:        {grid2.best_score_}')
# ==========================================================================
pipe3 = Pipeline([('tv', TfidfVectorizer()),('lr', LogisticRegression())])

pipe3_params = {'tv__max_features': [100,500],
                'tv__stop_words': ['english'],
                'tv__ngram_range': [(1,1),(1,2)],
              } 

grid3 = GridSearchCV(pipe3,pipe3_params,cv=5)
grid3.fit(X_train, y_train)
print(f'TfidVectorizer+LogisticRegression:  {grid3.best_score_}')
# ==========================================================================
pipe4 = Pipeline([('tv', TfidfVectorizer()),('mb', MultinomialNB())])

pipe4_params = {'tv__max_features': [100,500],
                'tv__stop_words': ['english'],
                'tv__ngram_range': [(1,1),(1,2)],
              } 

grid4 = GridSearchCV(pipe4,pipe4_params,cv=5)
grid4.fit(X_train, y_train)
print(f'TfidVectorizer+NaieveBayes:         {grid4.best_score_}')

CountVectorizer+LogisticRegression: 0.5438682484651949
CountVectorizer+NaieveBayes:        0.5381053896091771
TfidVectorizer+LogisticRegression:  0.5455964214505491
TfidVectorizer+NaieveBayes:         0.5473257355549884


**10. What are some of the advantages of fitting a logistic regression model?**

- First advantage of all time "fairly easy to understand and explain"
- a great starter algorithm for text related classification

**11. Fit a logistic regression model and compare it to the baseline.**

In [9]:
y_test.value_counts(normalize=True).mul(100).round(2)

y    53.11
n    46.89
Name: cAGR, dtype: float64

The baseline score and LogisticRegression gave out quite similar score.

### Summary of Naive Bayes 

Naive Bayes is a classification technique that relies on probability to classify observations.
- It's based on a probability rule called **Bayes' Theorem**... thus, "**Bayes**."
- It makes an assumption that isn't often met, so it's "**naive**."

Despite being a model that relies on a naive assumption, it often performs pretty well! (This is kind of like linear regression... we aren't always guaranteed homoscedastic errors in linear regression, but the model might still do a good job regardless.)
- [Interested in details? Read more here if you want.](https://www.cs.unb.ca/~hzhang/publications/FLAIRS04ZhangH.pdf)


The [sklearn documentation](https://scikit-learn.org/stable/modules/naive_bayes.html) is here, but it can be intimidating. So, to quickly summarize the Bayes and Naive parts of the model...

#### Bayes' Theorem
If you've seen Bayes' Theorem, it relates the probability of $P(A|B)$ to $P(B|A)$. (Don't worry; we won't be doing any probability calculations by hand! However, you may want to refresh your memory on conditional probability from our earlier lessons if you forget what a conditional probability is.)

$$
\begin{eqnarray*}
\text{Bayes' Theorem: } P(A|B) &=& \frac{P(B|A)P(A)}{P(B)}
\end{eqnarray*}
$$

- Let $A$ be that someone is "agreeable," like the OCEAN category.
- Let $B$ represent the words used in their Facebook post.

$$
\begin{eqnarray*}
\text{Bayes' Theorem: } P(A|B) &=& \frac{P(B|A)P(A)}{P(B)} \\
\Rightarrow P(\text{person is agreeable}|\text{words in Facebook post}) &=& \frac{P(\text{words in Facebook post}|\text{person is agreeable})P(\text{person is agreeable})}{P(\text{words in Facebook post})}
\end{eqnarray*}
$$

We want to calculate the probability that someone is agreeable **given** the words that they used in their Facebook post! (Rather than calculating this probability by hand, this is done under the hood and we can just see the results by checking `.predict_proba()`.) However, this is exactly what our model is doing. We can (a.k.a. the model can) calculate the pieces on the right-hand side of the equation to give us a probability estimate of how likely someone is to be agreeable given their Facebook post.

#### Naive Assumption

If our goal is to estimate $P(\text{person is agreeable}|\text{words in Facebook post})$, that can be quite tricky.

---

<details><summary>Bonus: if you want to understand why that's complicated, click here.</summary>
    
- The event $\text{"words in Facebook post"}$ is a complicated event to calculate.

- If a Facebook post has 100 words in it, then the event $\text{"words in Facebook post"} = \text{"word 1 is in the Facebook post" and "word 2 is in the Facebook post" and }\ldots \text{ and "word 100 is in the Facebook post"}$.

- To calculate the joint probability of all 100 words being in the Facebook post gets complicated pretty quickly. (Refer back to the probability notes on how to calculate the joint probability of two events if you want to see more.)
</details>

---

To simplify matters, we make an assumption: **we assume that all of our features are independent of one another.**

In some contexts, this assumption might be realistic!

**12. Why would this assumption not be realistic with NLP data?**

After tokenizing words, context might be loss and make it harder for assumption 

---

Despite this assumption not being realistic with NLP data, we still use Naive Bayes pretty frequently.
- It's a very fast modeling algorithm. (which is great especially when we have lots of features and/or lots of data!)
- It is often an excellent classifier, outperforming more complicated models.

There are three common types of Naive Bayes models: Bernoulli Naive Bayes, Multinomial Naive Bayes, and Gaussian Naive Bayes.
- How do we pick which of the three models to use? It depends on our $X$ variable.
    - Bernoulli Naive Bayes is appropriate when our features are all 0/1 variables.
        - [Bernoulli NB Documentation](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.BernoulliNB.html#sklearn.naive_bayes.BernoulliNB)
    - Multinomial Naive Bayes is appropriate when our features are variables that take on only positive integer counts.
        - [Multinomial NB Documentation](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html#sklearn.naive_bayes.MultinomialNB)
    - Gaussian Naive Bayes is appropriate when our features are Normally distributed variables. (Realistically, though, we kind of use Gaussian whenever neither Bernoulli nor Multinomial works.)
        - [Gaussian NB Documentation](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html#sklearn.naive_bayes.GaussianNB)

---

**13. Suppose you CountVectorized your features. Which Naive Bayes model would be most appropriate to fit? Why? Fit it.**

- Multinomial Naive Bayes is the most appropriate because words are tokenized into many variables
- not [0,1] in Bernoulli type
- and not Gaussian type for Normal distribution

**14. Suppose you TFIDFVectorized your features. Which Naive Bayes model would be most appropriate to fit? Why? Fit it.**

same answer as the above

**15. Compare the performance of your models.**

Parameters are max_features = 100,500 --> ngram_range = (1,1),(1,2) --> stop_words = english 
- CountVectorizer+LogisticRegression: 0.5438682484651949
- CountVectorizer+NaieveBayes:        0.5381053896091771
- TfidVectorizer+LogisticRegression:  0.5455964214505491
- TfidVectorizer+NaieveBayes:         0.5473257355549884

**16. Even though we didn't explore the full extent of Cambridge Analytica's modeling, based on what we did here, how effective was their approach at using Facebook data to model agreeableness?**

Baseline score and Model score can't distinguish Agreeableness personality that much because they are around 50/50 chance